Enable experimental feature to support tab completion on abbreviated cmdlets

SteveL-MSFT opened the pull request:
## PR Summary

While watching some demos at PSConfAsia, it's obvious that cmdlet nouns have gotten descriptive and wordy (Azure and AWS are good examples).  This `experimental` feature enables tab-completion of abbreviated cmdlets and functions:

```none
i-psdf<tab>

returns: Import-PowerShellDataFile

u-akvmssdr<tab>

returns: Undo-AzKeyVaultManagedStorageSasDefinitionRemoval
```

Note that this only works for tab completion (interactive use), so `i-psdf` is still an invalid cmdlet name in scripts.

The matching is entirely based on the capital letters of the verb and noun.  To test this, create test.json:

```json
{
  "ExperimentalFeatures": [
    "PSUseAbbreviationExpansion"
  ]
}
```

And run:

> pwsh -settingsfile ./test.json

Associated [RFC](https://github.com/PowerShell/PowerShell-RFC/pull/150/files)

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

SteveL-MSFT committed:
[feature]
enable experimental feature to support tab completion on abbreviations of cmdlets

SteveL-MSFT committed:
[feature]
address codefactor issues

anmenaga commented:
This needs another `[feature]` test pass to make sure it is green.

anmenaga commented:
@daxian-dbw @BrucePay This PR needs your review.

JamesWTruher commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -159,28 +136,74 @@ private static List<CompletionResult> CompleteCommand(CompletionContext context,
+                    powershell.AddParameter("CommandType", types);
+                }
+
+                commandInfos = context.Helper.ExecuteCurrentPowerShell(out exceptionThrown);
```
shouldn't we do something if there was an error?

JamesWTruher commented on src/System.Management.Automation/engine/CommandSearcher.cs:
```diff
@@ -978,6 +988,14 @@ private CmdletInfo GetNextCmdlet()
+                            else if (_commandResolutionOptions.HasFlag(SearchResolutionOptions.UseAbbreviationExpansion))
+                            {
+                                string abbreviatedCmdlet = new string(cmdlet.Name.Where(c => char.IsUpper(c) || c == '-').ToArray());
+                                if (_commandName.Equals(abbreviatedCmdlet, StringComparison.OrdinalIgnoreCase))
```
I wonder if this should be `abbreviatedCmdlet.IndexOf(_commandName, StringComparison.OrdinalIgnoreCase) == 0`
which would allow `pschr` to match `PSConsoleHostReadLine`

SteveL-MSFT replied:
The current code already resolves `pschr` as `PSConsoleHostReadLine`.  That line is so that `i-psdf` resolves `Import-PowerShellDataFile`

JamesWTruher replied:
see my response below

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1:
```diff
@@ -0,0 +1,68 @@
+        }
+
+        It "Can return multiple results for cmdlets matching abbreviation" {
+            $results = pwsh -settingsfile $configFilePath -c "Get-Command i-C -UseAbbreviationExpansion | ConvertTo-Json" | ConvertFrom-Json
```
was `i-C` casing done purposefully?

SteveL-MSFT replied:
Yes, to test case-insensitivity

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1:
```diff
@@ -0,0 +1,68 @@
+        }
+
+        It "No results if wildcard is used" {
+            pwsh -settingsfile $configFilePath -c Get-Command i-psd* -UseAbbreviationExpansion | Should -BeNullOrEmpty
```
my assumption is that if there is a command `i-psdf` then this would be found (due to the usual path for get command), is that right? `-UseAbbreviationExpansion` only adds an additional hunt for a command?

JamesWTruher replied:
i suppose that indicates that we should have an additional test to ensure that `-UseAbbreviationExpansion` doesn't negatively effect command lookup

SteveL-MSFT replied:
The intent of `UseAbbreviationExpansion` is for exact match only (used as an additional hunt with the existing searching logic).  So `i-psdf` will resolve, but `i-psd*` is nonsense.  Can you give me an example of what the additional test would look like?  Not sure I understand.

JamesWTruher replied:
with regard to our current tab completion, a partial match provides all possibilities, so `Get-C<tab>` will match `get-content`, `get-childitem`, etc. It seems to me that we have *increased* the cognitive burden with this change because now the user has to remember how the cmdlet name has been cased which is arbitrary and capricious (IMO). Supporting my example above would allow any cmdlet which case matches `I*-P*S*D*` seems quite reasonable to me since I can't remember how we case things. The more compound the noun the higher the burden.

SteveL-MSFT replied:
This change doesn't affect the current tab-completion behavior, so `Get-C<tab>` still works as it works today.  The intent of this change isn't to provide another way to search for commands, but to allow tab completion of a shortcut to a well known command.  If there's a desire to support `i-<tab>` as a way to find all commands where the verb starts with `i`, that should be feedback to the associated RFC.

JamesWTruher reviewed (approved):
there's nothing in my comments which are blocking, but I *am* worried about the case where we have an error (CompletionCompleters, line 193)

SteveL-MSFT committed:
address Jim's feedback

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -159,28 +136,75 @@ private static List<CompletionResult> CompleteCommand(CompletionContext context,
 
-                    Exception exceptionThrown;
-                    var commandInfos = context.Helper.ExecuteCurrentPowerShell(out exceptionThrown);
+        private static List<CompletionResult> ExecuteGetCommandCommand(CompletionContext context, string moduleName, CommandTypes types, bool includeModulePrefix)
```
I recommend to have this method as a local function within `CompleteCommand`. This method recalculate `addAmpersandIfNecessary` and `quote`. We can avoid the duplication if we use a local function.

SteveL-MSFT replied:
will do

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -159,28 +136,75 @@ private static List<CompletionResult> CompleteCommand(CompletionContext context,
+                    .AddParameter("All")
+                    .AddParameter("Name", commandName);
+
+                    if (ExperimentalFeature.IsEnabled("PSUseAbbreviationExpansion"))
```
Indentation should be fixed.

SteveL-MSFT replied:
Will fix

daxian-dbw commented on src/System.Management.Automation/engine/CommandDiscovery.cs:
```diff
@@ -1502,7 +1520,24 @@ internal IEnumerator<CmdletInfo> GetCmdletInfo(string cmdletName, bool searchAll
+                        {
+                            foreach (CmdletInfo cmdletInfo in cmdletList)
+                            {
+                                string abbreviatedCmdlet = new string(cmdletInfo.Name.Where(c => char.IsUpper(c)).ToArray());
```
Do we care about `-` here?

SteveL-MSFT replied:
Actually, this code is no longer needed since I'm treating `-UseAbbreviationExpansion` as a command name pattern so the code will never call this method when `-UseAbbreviationExpansion` is used.  I'll clean this up.

SteveL-MSFT commented on src/System.Management.Automation/engine/CommandDiscovery.cs:
```diff
@@ -1131,11 +1131,27 @@ private static CommandInfo InvokeCommandNotFoundHandler(string commandName, Exec
+                                    string abbreviatedCmdlet = new string(cmdlet.Key.Where(c => char.IsUpper(c) || c == '-').ToArray());
+                                    if (commandName.Equals(abbreviatedCmdlet, StringComparison.OrdinalIgnoreCase))
+                                    {
+                                        result = LookupCommandInfo(cmdlet.Key, commandOrigin, context);
```
LookupCommandInfo() does automoduleloading (line 828 in commanddiscovery.cs)

daxian-dbw replied:
`TryModuleAutoLoading` seems only loading the module if the command name is fully qualified. `LookupCommandInfo` depends on `TryModuleAutoDiscovery` to load the module if the auto-discovery finds anything, and that is supposed to happen in `if (exportedCommands.TryGetValue(commandName, out exportedCommandTypes))` via the call to `AutoloadSpecifiedModule`.

Your current implementation actually does 2-pass when the command is discovered from a module:
- in the first pass, you only resolve the abbreviated command name to the full command name, and then call `LookupCommandInfo` with the resolved command name, which will start the 2nd auto-discovery pass because the module is not yet loaded.
- in the second pass, the execution will get in `if (exportedCommands.TryGetValue(commandName, out exportedCommandTypes))`, and the module is loaded via `AutoloadSpecifiedModule`.

SteveL-MSFT replied:
I believe this is addressed with the additional code change and test added for auto-loading

SteveL-MSFT commented on src/System.Management.Automation/engine/CommandDiscovery.cs:
```diff
@@ -1131,11 +1131,27 @@ private static CommandInfo InvokeCommandNotFoundHandler(string commandName, Exec
+                                foreach (var cmdlet in exportedCommands)
+                                {
+                                    string abbreviatedCmdlet = new string(cmdlet.Key.Where(c => char.IsUpper(c) || c == '-').ToArray());
+                                    if (commandName.Equals(abbreviatedCmdlet, StringComparison.OrdinalIgnoreCase))
```
This is analogous to how this method works w/o this new code (see line 1152).  I can add a comment to make this more clear, however.

daxian-dbw replied:
Without the new code, we are looking for the command name from all commands exposed from a module, and there won't be duplicate command names.
The new code looks for the abbreviated name, and it's possible, rare though, that two commands from the same module has the same abbreviated names.

SteveL-MSFT commented on src/System.Management.Automation/engine/CommandDiscovery.cs:
```diff
@@ -1131,11 +1131,27 @@ private static CommandInfo InvokeCommandNotFoundHandler(string commandName, Exec
+                                    break;
+                                }
                             }
 
```
Will change

SteveL-MSFT committed:
address Dongbo's feedback
fixed code to support returning multiple results and not just first one
added test for automoduleloading which also covers multiple results from different modules

SteveL-MSFT committed:
refactor ExecuteGetCommandCommand as local function

SteveL-MSFT commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -82,48 +82,88 @@ internal static List<CompletionResult> CompleteCommand(CompletionContext context
         private static List<CompletionResult> CompleteCommand(CompletionContext context, string moduleName, CommandTypes types = CommandTypes.All)
         {
-            var addAmpersandIfNecessary = IsAmpersandNeeded(context, false);
+            List<CompletionResult> ExecuteGetCommandCommand(string command, CompletionContext completionContext, string module, CommandTypes commandTypes, bool useModulePrefix, bool addAmpersand, string quotes)
```
ok, wasn't aware of that!  will fix.

SteveL-MSFT commented on src/System.Management.Automation/engine/CommandSearcher.cs:
```diff
@@ -1578,5 +1595,9 @@ internal enum SearchResolutionOptions
+        /// <summary>
+        /// Enable searching for cmdlets/functions by abbreviation expansion.
+        /// </summary>
+        UseAbbreviationExpansion = 0x10,
```
`CommandNameIsPattern` set only if the [command name contains wildcards](https://github.com/PowerShell/PowerShell/blob/d80154430d600a52a1d1e3a9f3ecf032b36185d4/src/System.Management.Automation/engine/GetCommandCommand.cs#L720).  In this case when using abbreviation expansion, the command name sans wild card still acts like a pattern.

daxian-dbw replied:
But you changed [that check](https://github.com/PowerShell/PowerShell/blob/d80154430d600a52a1d1e3a9f3ecf032b36185d4/src/System.Management.Automation/engine/GetCommandCommand.cs#L717) to `bool isPattern = WildcardPattern.ContainsWildcardCharacters(plainCommandName) || UseAbbreviationExpansion;`, which means `UseAbbreviationExpansion` implies `CommandNameIsPattern`.

SteveL-MSFT replied:
@daxian-dbw It's not that `UseAbbreviationExpansion` implies `ResolveFunctionPatterns`.  The default if `-CommandType` is not specified is [`All`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/GetCommandCommand.cs#L198) which [includes](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/GetCommandCommand.cs#L711) `ResolveFunctionPatterns`.  `CommandNameIsPattern` is part of the same enum (incorrectly in my opinion) and used for a different purpose.  It's to indicate that the commandName is a filter rather than literal.  Since `UseAbbreviationExpansion` is also being used as a filter, I added the line to treat it as `isPattern` so that it goes through the filtering (wildcard matching) code path.

SteveL-MSFT commented on src/System.Management.Automation/engine/CommandDiscovery.cs:
```diff
@@ -1131,12 +1131,33 @@ private static CommandInfo InvokeCommandNotFoundHandler(string commandName, Exec
+                                if (result != null)
+                                {
+                                    break;
+                                }
```
yes

SteveL-MSFT committed:
[feature]
address Dongbo's feedback

SteveL-MSFT commented on src/System.Management.Automation/engine/CommandDiscovery.cs:
```diff
@@ -1131,12 +1131,33 @@ private static CommandInfo InvokeCommandNotFoundHandler(string commandName, Exec
                             {
                                 break;
                             }
+                            else if (searchResolutionOptions.HasFlag(SearchResolutionOptions.UseAbbreviationExpansion))
```
Verified this is no longer needed

SteveL-MSFT commented on src/System.Management.Automation/engine/GetCommandCommand.cs:
```diff
@@ -913,7 +927,7 @@ private bool FindCommandForName(SearchResolutionOptions options, string commandN
                     // Only for this case, the loop should exit
                     // Get-Command Foo
-                    if (isPattern || All || TotalCount != -1 || _isCommandTypeSpecified || _isModuleSpecified || _isFullyQualifiedModuleSpecified)
+                    if (UseAbbreviationExpansion || isPattern || All || TotalCount != -1 || _isCommandTypeSpecified || _isModuleSpecified || _isFullyQualifiedModuleSpecified)
```
yes, removed

daxian-dbw commented on src/System.Management.Automation/engine/CommandSearcher.cs:
```diff
@@ -5,6 +5,8 @@
 using System.Collections.Generic;
 using System.Collections.ObjectModel;
 using System.IO;
+using System.Linq;
```
This using directive is can be removed now.

daxian-dbw commented on src/System.Management.Automation/engine/Modules/ModuleUtils.cs:
```diff
@@ -4,7 +4,9 @@
 using System.Collections.Generic;
 using System.Globalization;
 using System.IO;
+using System.Linq;
```
This using directive can be removed now.

daxian-dbw commented on src/System.Management.Automation/engine/Modules/ModuleUtils.cs:
```diff
@@ -556,6 +559,25 @@ internal static IEnumerable<CommandInfo> GetMatchingCommands(string pattern, Exe
+        /// <returns>Abbreviated version of the command name.</returns>
+        internal static string AbbreviateName(string commandName)
+        {
+            StringBuilder abbreviation = new StringBuilder();
```
nitpick: the default capacity of `StringBuilder` is 4, which means a 2-word noun part of the command name. Maybe use a default capacity, say 6? That means 4 words in the noun part of the command name, which should be large enough for most cases.
This is not a blocking comment.

daxian-dbw commented on src/System.Management.Automation/engine/GetCommandCommand.cs:
```diff
@@ -742,7 +756,7 @@ private void AccumulateMatchingCommands(IEnumerable<string> commandNames)
                             try
                             {
-                                CommandDiscovery.LookupCommandInfo(tempCommandName, this.MyInvocation.CommandOrigin, this.Context);
+                                CommandDiscovery.LookupCommandInfo(tempCommandName, CommandTypes.All, options, this.MyInvocation.CommandOrigin, this.Context);
```
If we don't care about the case that `isPattern == true` (`-abbreviation` specified) and `moudleName` is not-null-or-empty, then we don't need to change this. In that case, the execution will fall in `TryAutoloadingModule`, which doesn't handle pattern at all 😄

SteveL-MSFT committed:
address Dongbo's feedback

daxian-dbw reviewed (approved):
LGTM
